Add usage stats for semantic_text fields

dimitris-athanasiou opened the pull request:
This change enhances usage reporting for the inference plugin to account for usage of `semantic_text` fields.

Usage is bucketed by service and task_type. For each bucket this adds a `semantic_text` object that contains:

- `field_count`: the number of `semantic_text` fields that use an inference endpoint of that service/task_type.
- `indices_count`: the number of indices that contain at least one `semantic_field` referencing an inference endpoint of that service/task_type.
- `inference_id_count`: the number of distinct inference endpoints of that service/task_type used by `semantic_text` fields.

In addition, this change adds two new kinds of buckets that facalitate getting aggregate usage information:

- `_all` buckets are added by `task_type`. These allow summation of usage info for all inference endpoints of a particular `task_type`.
- default model buckets are added. Those contain usage info for models that are included by default.

dimitris-athanasiou committed:
Add usage stats for semantic_text fields

This change enhances usage reporting for the inference plugin
to account for usage of `semantic_text` fields.

Usage is bucketed by service and task_type. For each bucket
this adds a `semantic_text` object that contains:

- `field_count`: the number of `semantic_text` fields that use
an inference endpoint of that service/task_type.
- `indices_count`: the number of indices that contain at least
one `semantic_field` referencing an inference endpoint of that service/task_type.
- `inference_id_count`: the number of distinct inference endpoints
of that service/task_type used by `semantic_text` fields.

In addition, this change adds two new kinds of buckets that facalitate
getting aggregate usage information:

- `_all` buckets are added by `task_type`. These allow summation of usage
info for all inference endpoints of a particular `task_type`.
- default model buckets are added. Those contain usage info for models
that are included by default.

dimitris-athanasiou committed:
Update docs/changelog/135262.yaml

dimitris-athanasiou committed:
Fix changelog

kderusso reviewed (commented):
Nice work! Are there any existing yaml tests that we could update here too?

dimitris-athanasiou committed:
Prepare SemanticTextStatsTests for BWC testing

dimitris-athanasiou committed:
Add YAML test

dimitris-athanasiou commented:
I have added YAML tests as per your suggestion @kderusso

dimitris-athanasiou committed:
Fix YAML test

dimitris-athanasiou committed:
Revert "Fix YAML test"

This reverts commit d72707d9311a2e06689edb4098a84422a604ec1d.

dimitris-athanasiou committed:
Strip linux suffix from model_id for default stats

dimitris-athanasiou commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/action/TransportInferenceUsageAction.java:
```diff
@@ -62,17 +79,171 @@ protected void localClusterStateOperation(
+        @Override
+        public String toString() {
+            // Inference ids cannot start with '_'. Thus, default stats do to avoid conflicts with user-defined inference ids.
+            return "_" + service + "_" + modelId.replace('.', '_');
```
The API validates that the inference id starts with `[a-z0-9]`. It has no knowledge of how we go about to create keys for usage stats. I might be misunderstanding something but the way it currently is, we'd have to add extra validation code when an inference endpoint is created to avoid conflicts with the way we're naming the default stats buckets.

kderusso replied:
That's correct. The default endpoints all start with `.` so they would fail existing validation.

I think this a moot point at this point, we're agreeing on validation but talking past each other. 🙂

dimitris-athanasiou committed:
Correct linux suffix this time

Mikep86 commented on x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/inference/inference_usage.yml:
```diff
@@ -0,0 +1,146 @@
+setup:
+  - requires:
+      cluster_features: "inference.semantic_text_usage"
+      reason: tests added after semantic_text usage was added
```
Nit: It's a bit redundant to mention tests in the reason here. We could say something like "semantic text stats added to inference usage" here to be clearer.

Mikep86 commented on x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/inference/inference_usage.yml:
```diff
@@ -0,0 +1,146 @@
+  - match: { inference.models.0.service: "_all" }
+  - match: { inference.models.0.task_type: "RERANK" }
+  - match: { inference.models.0.count: 1 }
+  - is_false: inference.models.0.semantic_text
```
Nit: the `not_exists` assertion does basically the same thing, but the intent is clearer

Mikep86 commented on x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/inference/inference_usage.yml:
```diff
@@ -0,0 +1,146 @@
+                type: semantic_text
+              field_2:
+                type: semantic_text
+                inference_id: .multilingual-e5-small-elasticsearch
```
A note on default endpoint usage in CI: We have noticed flakiness when we actually use them to generate embeddings due to deployment timeouts and failures. This reference should be fine because it's not actually triggering deployment of the E5 model, but something to be aware of in general.

Mikep86 reviewed (commented):
Great work! I left some comments, but they're almost all nits. Per our offline discussion, I think we need to figure out if we want to filter out all hidden indices (i.e. those that start with `.`). Other than that, this PR is in good shape 🚀

dimitris-athanasiou committed:
Changelog area is Vector Search

dimitris-athanasiou commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/action/TransportInferenceUsageAction.java:
```diff
@@ -62,17 +79,181 @@ protected void localClusterStateOperation(
+        );
+    }
+
+    private static void addTopLevelSemanticTextStatsByTask(
```
Good point!

dimitris-athanasiou committed:
Address some review points

dimitris-athanasiou committed:
Address evil edge case

dimitris-athanasiou commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/action/TransportInferenceUsageAction.java:
```diff
@@ -62,17 +79,181 @@ protected void localClusterStateOperation(
+                Map.of()
+            );
+            fieldsByIndex = filterFields(fieldsByIndex, f -> statKey.modelId.equals(endpointIdToModelId.get(f.getInferenceId())));
+            ModelStats stats = new ModelStats(statKey.toString(), statKey.taskType, defaultModelStatsKeyToEndpointCount.getValue());
+            addSemanticTextStats(fieldsByIndex, stats);
+            endpointStats.put(statKey.toString(), stats);
```
This would take some evil doing but it is possible! I have addressed it in https://github.com/elastic/elasticsearch/pull/135262/commits/4f7955890ace9d1e2ed3e8ed6975414cff762f9b

dimitris-athanasiou committed:
Do not omit zero values

Commit:
[CI] Auto commit changes from spotless

dimitris-athanasiou committed:
Also exclude hidden indices

dimitris-athanasiou commented:
@Mikep86 Regarding hidden indices. I have also excluded them for now. They are probably of little value telemetry-wise. However, if we want to collect them in the future we can easily have a separate section for them by adding a `SemanticTextStats` object as a member of `SemanticTextStats`. We could then gather system/hidden field stats in there. Had an offline chat with @jimczi about this solution and we're on the same page.

Mikep86 reviewed (commented):
Looks really good, thanks for the changes! Functionality looks solid. I left a few comments for one more round of small changes, then I think we're good to merge!

Mikep86 reviewed (approved):
LGTM! I have some minor comments in my previous review, but none of them are blocking.

dimitris-athanasiou committed:
Address more review comments

dimitris-athanasiou commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/inference/InferenceFeatureSetUsage.java:
```diff
@@ -66,7 +66,7 @@ public int hashCode() {
     }
 
-    Collection<ModelStats> modelStats() {
+    public Collection<ModelStats> modelStats() {
```
Made package private.

dimitris-athanasiou commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/action/TransportInferenceUsageAction.java:
```diff
@@ -62,17 +82,188 @@ protected void localClusterStateOperation(
+            return modelId.substring(0, modelId.length() - MODEL_ID_LINUX_SUFFIX.length());
+        }
+        return modelId;
+    }
```
Well spotted! I first thought of making it handle `null`, then decided to ensure I've filtered nulls before calling it. Those annotations were left behind!

dimitris-athanasiou commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/action/TransportInferenceUsageAction.java:
Agreed. Added some commentary to help our future selves amongst others :-)

dimitris-athanasiou commented:
@Mikep86 Thank you for additional review! I have addressed those points in https://github.com/elastic/elasticsearch/pull/135262/commits/3d4f273f31110c5138493560582e9c5237dae9a8

dimitris-athanasiou committed:
Merge branch 'main' into usage-for-semantic-text

dimitris-athanasiou committed:
Merge branch 'main' into usage-for-semantic-text

kderusso reviewed (approved):
Nice work!

davidkyle commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/action/TransportInferenceUsageAction.java:
```diff
@@ -62,17 +81,207 @@ protected void localClusterStateOperation(
+        Map<String, ModelStats> endpointStats
+    ) {
+        for (TaskType taskType : TaskType.values()) {
+            if (taskType == TaskType.ANY) {
```
The `RERANK`, `COMPLETION` and `CHAT_COMPLETION` task types will never appear in a semantic text field and also can be skipped here. 

Add a static function `TaskType [] semanticTextTaskTypes( return SPARSE + TEXT)` to TaskType and just iterate those.

dimitris-athanasiou replied:
At this point we are adding top level `_all` stats buckets. Even though we won't have `semantic_text_stats` for those tasks, we can still have top level buckets for `count` and for the consistency/symmetry with what we do for the semantic_text supporting tasks. What do you reckon?

davidkyle commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/action/TransportInferenceUsageAction.java:
```diff
@@ -62,17 +81,207 @@ protected void localClusterStateOperation(
+            .filter(endpoint -> endpoint.getServiceSettings().modelId() != null)
+            .map(endpoint -> stripLinuxSuffix(endpoint.getServiceSettings().modelId()))
+            .collect(Collectors.toSet());
+        return endpoints.stream()
+            .filter(endpoint -> endpoint.getServiceSettings().modelId() != null)
+            .filter(endpoint -> modelIds.contains(stripLinuxSuffix(endpoint.getServiceSettings().modelId())))
+            .map(
+                endpoint -> new DefaultModelStatsKey(
+                    endpoint.getService(),
+                    endpoint.getTaskType(),
+                    stripLinuxSuffix(endpoint.getServiceSettings().modelId())
+                )
+            )
```
Can you delete the stream to build the modelIds set above and include the `modelRegistry.containsDefaultConfigId(...` filter here. 

Basically, is this equivalent? 

```suggestion
        return endpoints.stream()
            .filter(endpoint -> endpoint.getServiceSettings().modelId() != null)
            .filter(endpoint -> modelRegistry.containsDefaultConfigId(endpoint.getInferenceEntityId()))
            .map(
                endpoint -> new DefaultModelStatsKey(
                    endpoint.getService(),
                    endpoint.getTaskType(),
                    stripLinuxSuffix(endpoint.getServiceSettings().modelId())
                )
            )
```

davidkyle commented on x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/inference/inference_usage.yml:
```diff
@@ -0,0 +1,155 @@
+  - match: { inference.models.0.service: "_all" }
+  - match: { inference.models.0.task_type: "CHAT_COMPLETION" }
+  - match: { inference.models.0.count: 0 }
+  - match: { inference.models.0.semantic_text: { field_count: 0, indices_count: 0, inference_id_count: 0} }
```
Semantic text is not compatible with the `CHAT_COMPLETION` task type. Should the semantic_text field be optional in these stats objects and not reported for incompatible task types

dimitris-athanasiou replied:
I have addressed this in https://github.com/elastic/elasticsearch/pull/135262/commits/809631b4946fb98c06f85f5d0e3a941fc279df1d

dimitris-athanasiou committed:
Merge branch 'main' into usage-for-semantic-text

dimitris-athanasiou committed:
Only add semantic_text stats if task_type is compatible

Mikep86 commented on server/src/main/java/org/elasticsearch/inference/TaskType.java:
```diff
@@ -52,6 +52,16 @@ public static TaskType fromStringOrStatusException(String name) {
         }
     }
 
+    private final boolean isCompatibleWithSemanticText;
+
+    TaskType(boolean isCompatibleWithSemanticText) {
+        this.isCompatibleWithSemanticText = isCompatibleWithSemanticText;
+    }
```
I personally wouldn't modify `TaskType` directly to add this metadata. We only need it in `TransportInferenceUsageAction`, I would build a set of task types that need semantic text stats in that class.

Mikep86 commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/inference/usage/ModelStats.java:
```diff
@@ -19,28 +22,42 @@
     }
 
     public ModelStats(String service, TaskType taskType, long count) {
+        this(service, taskType, count, taskType.isCompatibleWithSemanticText() ? new SemanticTextStats() : null);
```
I think it would be simpler to always set `semanticTextStats` to null by default. Code paths in `TransportInferenceUsageAction` can pass in a non-null instance as necessary.

Mikep86 reviewed (commented):
I have some suggestions about how to simplify the implementation

dimitris-athanasiou committed:
Contain task type compatibility in TransportInferenceUsageAction

dimitris-athanasiou committed:
Merge branch 'main' into usage-for-semantic-text

Mikep86 reviewed (approved):
LGTM! Thanks for the continued iterations on this :)

dimitris-athanasiou committed:
Merge branch 'main' into usage-for-semantic-text

davidkyle reviewed (approved):
LGTM
